File: designs/creationalPatterns/prototype/prototype.ts
class Person {
  constructor(public name: string, public age: number) {}

  clone(): this {
    // Clona el objeto actual
    const clone = Object.create(this);
    return clone;
  }
}

// Crear un prototipo de persona
const personPrototype = new Person('John', 30);

// Clonar el prototipo para crear objetos Employee y Customer
const employee = personPrototype.clone();
employee.name = 'Employee';
employee.age = 25;

const customer = personPrototype.clone();
customer.name = 'Customer';
customer.age = 40;

// Mostrar los objetos clonados
console.log(employee); // Output: Employee { name: 'Employee', age: 25 }
console.log(customer); // Output: Customer { name: 'Customer', age: 40 }

File: src/creationalPatterns/prototype/prototype.py
import copy

class Person:
    def __init__(self, name, age):
        self.name = name
        self.age = age

    def clone(self):
        # Utilizamos la función deepcopy de la biblioteca copy para realizar una clonación profunda
        return copy.deepcopy(self)

# Crear un prototipo de persona
person_prototype = Person("John", 30)

# Clonar el prototipo para crear objetos Employee y Customer personalizados
employee = person_prototype.clone()
employee.name = "Employee"
employee.age = 25

customer = person_prototype.clone()
customer.name = "Customer"
customer.age = 40

# Mostrar los objetos clonados
print("Objeto Employee:", employee.__dict__)
print("Objeto Customer:", customer.__dict__)
